pkg/io/video/scale.go

```diff
@@ -0,0 +1,175 @@
+		}
+	}
+
+	return func(r Reader) Reader {
+		var rect image.Rectangle
+		var imgScaled, imgScaledCopy image.Image
+		if width > 0 && height > 0 {
+			rect = image.Rect(0, 0, width, height)
+		} else if width <= 0 && height <= 0 {
+			panic("Both width and height are negative!")
+		}
+
+		src := &rgbLikeYCbCr{y: &image.Gray{}, cb: &image.Gray{}, cr: &image.Gray{}}
+		dst := &rgbLikeYCbCr{y: &image.Gray{}, cb: &image.Gray{}, cr: &image.Gray{}}
+
+		// fixedRect returns Rectangle of chroma plane
+		fixedRect := func(rect image.Rectangle, sr image.YCbCrSubsampleRatio) image.Rectangle {
+			switch sr {
+			case image.YCbCrSubsampleRatio444:
+			case image.YCbCrSubsampleRatio422:
+				rect.Max.X /= 2
+			case image.YCbCrSubsampleRatio420:
+				rect.Max.X /= 2
+				rect.Max.Y /= 2
+			}
+			return rect
+		}
+		// ycbcrRealloc reallocs image.YCbCr if needed
+		ycbcrRealloc := func(i1 *image.YCbCr) {
+			if imgScaled == nil {
+				imgScaled = image.NewYCbCr(rect, i1.SubsampleRatio)
+				imgScaledCopy = image.NewYCbCr(rect, i1.SubsampleRatio)
+			}
+			imgDst := imgScaled.(*image.YCbCr)
+			yDx := rect.Dx()
+			yDy := rect.Dy()
+			cRect := fixedRect(rect, i1.SubsampleRatio)
+			cDx := cRect.Dx()
+			cDy := cRect.Dx()
+			yLen := yDx * yDy
+			cLen := cDx * cDy
+			if len(imgDst.Y) < yLen {
+				if cap(imgDst.Y) < yLen {
+					imgDst.Y = make([]uint8, yLen)
+				}
+				imgDst.Y = imgDst.Y[:yLen]
+			}
+			if len(imgDst.Cr) < cLen {
+				if cap(imgDst.Cr) < cLen {
+					imgDst.Cr = make([]uint8, cLen)
+				}
+				imgDst.Cr = imgDst.Cr[:cLen]
+			}
+			if len(imgDst.Cb) < cLen {
+				if cap(imgDst.Cb) < cLen {
+					imgDst.Cb = make([]uint8, cLen)
+				}
+				imgDst.Cb = imgDst.Cb[:cLen]
+			}
+			*dst.y = image.Gray{Pix: imgDst.Y, Stride: imgDst.YStride, Rect: rect}
+			*dst.cb = image.Gray{Pix: imgDst.Cb, Stride: imgDst.CStride, Rect: cRect}
+			*dst.cr = image.Gray{Pix: imgDst.Cr, Stride: imgDst.CStride, Rect: cRect}
+		}
+		// ycbcrRealloc reallocs image.RGBA if needed
+		rgbaRealloc := func(i1 *image.RGBA) {
+			if imgScaled == nil {
+				imgScaled = image.NewRGBA(rect)
+				imgScaledCopy = image.NewRGBA(rect)
+			}
+			imgDst := imgScaled.(*image.RGBA)
+			l := 4 * rect.Dx() * rect.Dy()
+			if len(imgDst.Pix) < l {
+				if cap(imgDst.Pix) < l {
+					imgDst.Pix = make([]uint8, l)
+				}
+				imgDst.Pix = imgDst.Pix[:l]
+			}
+		}
+
+		return ReaderFunc(func() (image.Image, error) {
+			img, err := r.Read()
+			if err != nil {
+				return nil, err
+			}
+
+			if imgScaled == nil {
+				if height <= 0 {
+					h := img.Bounds().Dy() * width / img.Bounds().Dx()
+					rect = image.Rect(0, 0, width, h)
+				} else if width <= 0 {
+					w := img.Bounds().Dx() * height / img.Bounds().Dy()
+					rect = image.Rect(0, 0, w, height)
+				}
+				cacheScaler(rect, img.Bounds())
+			}
+
+			switch v := img.(type) {
+			case *image.RGBA:
+				rgbaRealloc(v)
+				dst := imgScaled.(*image.RGBA)
```

Comment:
Not sure if this is going to happen. But, technically, this can cause a panic if suddenly the stream emits RGBA then later emits YCbCr. What do you think?

Reply:
Right, it and also https://github.com/pion/mediadevices/pull/55/files#r379957449 can happen. 
I'll put format change handling to `rgbaRealloc/ycbcrRealloc`.